In [15]:
# Step 1: Load and inspect the dataset
import pandas as pd

file_path = '/content/car_prices.csv'
df = pd.read_csv(file_path)

In [16]:
print(df.head(5))

   year   make                model        trim   body transmission  \
0  2015    Kia              Sorento          LX    SUV    automatic   
1  2015    Kia              Sorento          LX    SUV    automatic   
2  2014    BMW             3 Series  328i SULEV  Sedan    automatic   
3  2015  Volvo                  S60          T5  Sedan    automatic   
4  2014    BMW  6 Series Gran Coupe        650i  Sedan    automatic   

                 vin state  condition  odometer  color interior  \
0  5xyktca69fg566472    ca        5.0   16639.0  white    black   
1  5xyktca69fg561319    ca        5.0    9393.0  white    beige   
2  wba3c1c51ek116351    ca       45.0    1331.0   gray    black   
3  yv1612tb4f1310987    ca       41.0   14282.0  white    black   
4  wba6b2c57ed129731    ca       43.0    2641.0   gray    black   

                                   seller      mmr  sellingprice  \
0                 kia motors america  inc  20500.0       21500.0   
1                 kia motors ameri

In [17]:
df

year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558832  2015     Kia                 K900       Luxury      Sedan   
558833  2012     Ram                 2500  Power Wagon   Crew Cab   
558834  2012     BMW                   X5    xDrive35d        SUV   
558835  2015  Nissan               Altima        2.5 S      sedan   
558836  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer   color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0   white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0   white   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0    gray   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0   white   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0    gray   
...             ...                ...   ...        ...       ...     ...   
558832          NaN  knalw4d4xf6019304    in       45.0   18255.0  silver   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0   white   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0   black   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0   white   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0    gray   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558832    black                                   avis corporation  35300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice                                 saledate  
0            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2            30000.0  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3            27750.0  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4            67000.0  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)  
...              ...                                      ...  
558832       33000.0  Thu Jul 09 2015 07:00:00 GMT-0700 (PDT)  
558833       30800.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558834       34000.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558835       11100.0  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)  
558836       26700.0  Thu May 28 2015 05:30:00 GMT-0700 (PDT)  

[558837 rows x 16 columns]

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558837 entries, 0 to 558836
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558837 non-null  int64  
 1   make          548536 non-null  object 
 2   model         548438 non-null  object 
 3   trim          548186 non-null  object 
 4   body          545642 non-null  object 
 5   transmission  493485 non-null  object 
 6   vin           558833 non-null  object 
 7   state         558837 non-null  object 
 8   condition     547017 non-null  float64
 9   odometer      558743 non-null  float64
 10  color         558088 non-null  object 
 11  interior      558088 non-null  object 
 12  seller        558837 non-null  object 
 13  mmr           558799 non-null  float64
 14  sellingprice  558825 non-null  float64
 15  saledate      558825 non-null  object 
dtypes: float64(4), int64(1), object(11)
memory usage: 68.2+ MB


In [19]:
print(df.describe())


                year      condition       odometer            mmr  \
count  558837.000000  547017.000000  558743.000000  558799.000000   
mean     2010.038927      30.672365   68320.017767   13769.377495   
std         3.966864      13.402832   53398.542821    9679.967174   
min      1982.000000       1.000000       1.000000      25.000000   
25%      2007.000000      23.000000   28371.000000    7100.000000   
50%      2012.000000      35.000000   52254.000000   12250.000000   
75%      2013.000000      42.000000   99109.000000   18300.000000   
max      2015.000000      49.000000  999999.000000  182000.000000   

        sellingprice  
count  558825.000000  
mean    13611.358810  
std      9749.501628  
min         1.000000  
25%      6900.000000  
50%     12100.000000  
75%     18200.000000  
max    230000.000000  


In [20]:
# step 2 : handle missing values

print("\nMissing values in each column:")
print(df.isnull().sum())


Missing values in each column:
year                0
make            10301
model           10399
trim            10651
body            13195
transmission    65352
vin                 4
state               0
condition       11820
odometer           94
color             749
interior          749
seller              0
mmr                38
sellingprice       12
saledate           12
dtype: int64


In [21]:
df = df.dropna() #drop

In [22]:
print(df.isnull().sum())

year            0
make            0
model           0
trim            0
body            0
transmission    0
vin             0
state           0
condition       0
odometer        0
color           0
interior        0
seller          0
mmr             0
sellingprice    0
saledate        0
dtype: int64


In [23]:
df_cleaned_path = "cleaned_dataset.csv"
df.to_csv(df_cleaned_path, index=False)


In [24]:
df

year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558831  2011     BMW             5 Series         528i      Sedan   
558833  2012     Ram                 2500  Power Wagon   Crew Cab   
558834  2012     BMW                   X5    xDrive35d        SUV   
558835  2015  Nissan               Altima        2.5 S      sedan   
558836  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer  color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0  white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0  white   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0   gray   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0  white   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0   gray   
...             ...                ...   ...        ...       ...    ...   
558831    automatic  wbafr1c53bc744672    fl       39.0   66403.0  white   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0  white   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0  black   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0  white   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0   gray   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558831    brown           lauderdale imports ltd bmw pembrok pines  20300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice                                 saledate  
0            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2            30000.0  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3            27750.0  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4            67000.0  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)  
...              ...                                      ...  
558831       22800.0  Tue Jul 07 2015 06:15:00 GMT-0700 (PDT)  
558833       30800.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558834       34000.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558835       11100.0  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)  
558836       26700.0  Thu May 28 2015 05:30:00 GMT-0700 (PDT)  

[472325 rows x 16 columns]

In [25]:
#step 3: feature engineering

from datetime import datetime

# Create 'Car Age' column (current year - manufacturing year)
current_year = datetime.now().year
df['car_age'] = current_year - df['year']

# Create 'Price Difference' column (selling price - mmr)
df['price_difference'] = df['sellingprice'] - df['mmr']

<ipython-input-25-4e5b61459b17>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['car_age'] = current_year - df['year']
<ipython-input-25-4e5b61459b17>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_difference'] = df['sellingprice'] - df['mmr']


In [26]:
df

year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558831  2011     BMW             5 Series         528i      Sedan   
558833  2012     Ram                 2500  Power Wagon   Crew Cab   
558834  2012     BMW                   X5    xDrive35d        SUV   
558835  2015  Nissan               Altima        2.5 S      sedan   
558836  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer  color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0  white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0  white   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0   gray   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0  white   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0   gray   
...             ...                ...   ...        ...       ...    ...   
558831    automatic  wbafr1c53bc744672    fl       39.0   66403.0  white   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0  white   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0  black   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0  white   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0   gray   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558831    brown           lauderdale imports ltd bmw pembrok pines  20300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice                                 saledate  car_age  \
0            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)        9   
1            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)        9   
2            30000.0  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)       10   
3            27750.0  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)        9   
4            67000.0  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)       10   
...              ...                                      ...      ...   
558831       22800.0  Tue Jul 07 2015 06:15:00 GMT-0700 (PDT)       13   
558833       30800.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)       12   
558834       34000.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)       12   
558835       11100.0  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)        9   
558836       26700.0  Thu May 28 2015 05:30:00 GMT-0700 (PDT)       10   

        price_difference  
0                 1000.0  
1                  700.0  
2                -1900.0  
3                  250.0  
4                 1000.0  
...                  ...  
558831            2500.0  
558833             600.0  
558834            4200.0  
558835           -4000.0  
558836           -2900.0  

[472325 rows x 18 columns]

In [27]:
# Identify problematic rows in 'saledate'
print("Unique values in 'saledate' column (before fixing):")
print(df['saledate'].head())

# Convert 'saledate' to datetime format, forcing invalid dates to NaT
df['saledate'] = pd.to_datetime(df['saledate'], errors='coerce', utc=True)

# Display rows with invalid dates
invalid_dates = df[df['saledate'].isna()]
print("\nRows with invalid 'saledate' values:")
print(invalid_dates)

Unique values in 'saledate' column (before fixing):
0    Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
1    Tue Dec 16 2014 12:30:00 GMT-0800 (PST)
2    Thu Jan 15 2015 04:30:00 GMT-0800 (PST)
3    Thu Jan 29 2015 04:30:00 GMT-0800 (PST)
4    Thu Dec 18 2014 12:30:00 GMT-0800 (PST)
Name: saledate, dtype: object


<ipython-input-27-f6dc9d4a81e3>:6: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['saledate'] = pd.to_datetime(df['saledate'], errors='coerce', utc=True)



Rows with invalid 'saledate' values:
Empty DataFrame
Columns: [year, make, model, trim, body, transmission, vin, state, condition, odometer, color, interior, seller, mmr, sellingprice, saledate, car_age, price_difference]
Index: []


<ipython-input-27-f6dc9d4a81e3>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['saledate'] = pd.to_datetime(df['saledate'], errors='coerce', utc=True)


In [28]:
df = df.dropna(subset=['saledate'])

# Extract year and month from the corrected 'saledate'
df['sale_year'] = df['saledate'].dt.year
df['sale_month'] = df['saledate'].dt.month


In [29]:
# Inspect the new columns
print("\nSample of the new features:")
print(df[['year', 'car_age', 'sellingprice', 'mmr', 'price_difference', 'sale_year', 'sale_month']].head())


Sample of the new features:
   year  car_age  sellingprice      mmr  price_difference  sale_year  \
0  2015        9       21500.0  20500.0            1000.0       2014   
1  2015        9       21500.0  20800.0             700.0       2014   
2  2014       10       30000.0  31900.0           -1900.0       2015   
3  2015        9       27750.0  27500.0             250.0       2015   
4  2014       10       67000.0  66000.0            1000.0       2014   

   sale_month  
0          12  
1          12  
2           1  
3           1  
4          12  


In [30]:
# Save the dataset with new features
df_features_path = "car_prices_with_features.csv"
df.to_csv(df_features_path, index=False)
print(f"\nDataset with new features saved to {df_features_path}")



Dataset with new features saved to car_prices_with_features.csv


In [32]:
#step 4: ml model

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Step 1: Select relevant features for ML
ml_features = ['car_age', 'condition', 'odometer', 'price_difference', 'sale_year', 'sale_month']
target = 'sellingprice'


In [33]:
# Include categorical columns (e.g., make and state)
categorical_features = ['make', 'model', 'body', 'state']
ml_features += categorical_features

# Filter the dataset to include only selected features and drop rows with missing values in them
df_ml = df[ml_features + [target]].dropna()

In [34]:
# Step 2: Encode categorical variables
# Useed LabelEncoder for simplicity
label_encoders = {}
for col in categorical_features:
    le = LabelEncoder()
    df_ml[col] = le.fit_transform(df_ml[col].astype(str))
    label_encoders[col] = le

In [35]:
# Step 3: Split the data into training and testing sets
X = df_ml[ml_features]  # Features
y = df_ml[target]       # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [36]:
# Check the shapes of the resulting datasets
print(f"Training set shape: {X_train.shape}")
print(f"Testing set shape: {X_test.shape}")

Training set shape: (377860, 10)
Testing set shape: (94465, 10)


In [37]:
X_train.to_csv("X_train.csv", index=False)
X_test.to_csv("X_test.csv", index=False)
y_train.to_csv("y_train.csv", index=False)
y_test.to_csv("y_test.csv", index=False)
print("\nProcessed data saved!")



Processed data saved!


In [38]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [39]:
# Step 5: Train a Random Forest Regressor

rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

In [40]:
# Train the model on the training data
rf_model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [41]:
# Make predictions on the test set
y_pred = rf_model.predict(X_test)

In [43]:
# Evaluate the model

import numpy as np

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Model Performance:")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")

Model Performance:
Mean Absolute Error (MAE): 1367.67
Mean Squared Error (MSE): 6621093.98
Root Mean Squared Error (RMSE): 2573.15
R² Score: 0.93


In [44]:
import joblib
joblib.dump(rf_model, "random_forest_model.pkl")
print("\nRandom Forest Model saved as 'random_forest_model.pkl'")


Random Forest Model saved as 'random_forest_model.pkl'


Metrics Interpretation:
Mean Absolute Error (MAE): 1367.67

The average absolute error in predicted selling prices is $1,367.
For car prices, this is reasonable, especially if the dataset contains high-priced vehicles (e.g., luxury cars).
Root Mean Squared Error (RMSE): 2573.15

RMSE indicates the magnitude of typical prediction errors. A value of $2,573 means most predictions are within that range of the actual selling price.
R² Score: 0.93

Explains 93% of the variance in selling prices, which is an goof fit for a regression model.

In [45]:
df


year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558831  2011     BMW             5 Series         528i      Sedan   
558833  2012     Ram                 2500  Power Wagon   Crew Cab   
558834  2012     BMW                   X5    xDrive35d        SUV   
558835  2015  Nissan               Altima        2.5 S      sedan   
558836  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer  color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0  white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0  white   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0   gray   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0  white   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0   gray   
...             ...                ...   ...        ...       ...    ...   
558831    automatic  wbafr1c53bc744672    fl       39.0   66403.0  white   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0  white   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0  black   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0  white   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0   gray   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558831    brown           lauderdale imports ltd bmw pembrok pines  20300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice                  saledate  car_age  price_difference  \
0            21500.0 2014-12-16 04:30:00+00:00        9            1000.0   
1            21500.0 2014-12-16 04:30:00+00:00        9             700.0   
2            30000.0 2015-01-14 20:30:00+00:00       10           -1900.0   
3            27750.0 2015-01-28 20:30:00+00:00        9             250.0   
4            67000.0 2014-12-18 04:30:00+00:00       10            1000.0   
...              ...                       ...      ...               ...   
558831       22800.0 2015-07-06 23:15:00+00:00       13            2500.0   
558833       30800.0 2015-07-08 02:30:00+00:00       12             600.0   
558834       34000.0 2015-07-08 02:30:00+00:00       12            4200.0   
558835       11100.0 2015-07-08 23:45:00+00:00        9           -4000.0   
558836       26700.0 2015-05-27 22:30:00+00:00       10           -2900.0   

        sale_year  sale_month  
0            2014          12  
1            2014          12  
2            2015           1  
3            2015           1  
4            2014          12  
...           ...         ...  
558831       2015           7  
558833       2015           7  
558834       2015           7  
558835       2015           7  
558836       2015           5  

[472325